File: client/src/instruction/configUpdateAdmin.ts
import { Program } from "@project-serum/anchor";
import { PublicKey, TransactionInstruction } from "@solana/web3.js";
import { Cronos } from "../idl";
import { Account } from "../account";

export type ConfigUpdateAdminArgs = {
  newAdmin: PublicKey;
};

export class ConfigUpdateAdmin {
  private account: Account;
  private cronos: Program<Cronos>;

  constructor(account: Account, cronos: Program<Cronos>) {
    this.account = account;
    this.cronos = cronos;
  }

  public async configUpdateAdmin({
    newAdmin,
  }: ConfigUpdateAdminArgs): Promise<TransactionInstruction> {
    const configPDA = await this.account.config.pda();
    const configData = await this.account.config.data(configPDA.address);
    return this.cronos.instruction.configUpdateAdmin(newAdmin, {
      accounts: {
        admin: configData.admin,
        config: configPDA.address,
      },
    });
  }
}

File: src/instruction/config_update_admin.rs
use anchor_client::anchor_lang::{
    solana_program::{
        instruction::{AccountMeta, Instruction},
        pubkey::Pubkey,
    },
    InstructionData,
};

pub fn config_update_admin(admin: Pubkey, config: Pubkey, new_admin: Pubkey) -> Instruction {
    Instruction {
        program_id: cronos_program::ID,
        accounts: vec![
            AccountMeta::new(admin, true),
            AccountMeta::new(config, false),
        ],
        data: cronos_program::instruction::ConfigUpdateAdmin { new_admin }.data(),
    }
}
